zerver/tests/test_events.py

```diff
@@ -1478,6 +1480,12 @@ def test_do_reactivate_user(self) -> None:
         events = self.verify_action(action, num_events=2)
         check_realm_bot_add("events[1]", events[1])
 
+    def test_do_deactivate_realm(self) -> None:
+        realm = self.user_profile.realm
+        action = lambda: do_deactivate_realm(realm)
+        events = self.verify_action(action, state_change_expected=False)
```

Comment:
It's possible that this test would be more meaningful if you promoted Hamlet to an admin here, so that we can get an expected chance.

Reply:
Okay. I'll make the changes.

Is it to important to expect state change if a realm is deactivated?

Reply:
I think we want to at least understand why realm-deactivation doesn't change state.  If the event doesn't impact page_params (which is what the `state_change_expected` flag is indirectly related to), then it's questionable whether the app even needs this event.  I suspect you'll find a state change once you make Hamlet an admin.

A lot of this PR is gonna be about really digging into the code to understand what's happening here.

Reply:
Ok, I did a little more digging on my own here, and I think the next step for @shanukun here is to summarize my findings in the commit message.  So here is the code related to sending the event:

~~~
1013     event = dict(type="realm", op="deactivated", realm_id=realm.id)
1014     send_event(realm, event, active_user_ids(realm.id))
~~~

That code is pretty simple--it tells all the users in the realm that they have been deactivated.

And then the client code is pretty similar--we just redirect them to a deactivated page:

~~~
251             } else if (event.op === "deactivated") {
252                 window.location.href = "/accounts/deactivated/";
253             }
~~~

I was wondering why we don't have any apply_event code here, but the answer is super obvious to me now that I think about it--we don't send `page_params` to users who are deactivated, because, duh Steve, they can't run the app.

Given that, I don't think Shanu needs to modify the code here, except for a brief comment in the test that explains that `state_change_expected` is `False` here because realm deactivation is a unique situation.  And the challenge here is to try to summarize what I observe here a little more succinctly.  Tim -- this may be a situation where you're best off just modifying any comments/commit messages here to explain it more clearly.

Reply:
I think we actually probably do want `apply_event` code here; but that can be a follow-up task; I'll open an issue.